ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,330 @@
+                          self.load_simulated,
+                          self.load_true_images,
+                          self.load_trigger,
+                          self.load_instrument]
+        
+        self.subarray = SubarrayDescription.from_hdf(self.input_url)
+        self.h5file = tables.open_file(self.input_url, mode="r")
+
+        try:
+            self.structure = get_structure(self.h5file)
+        except ValueError:
+            self.structure = None
+
+        self.trigger_table = read_table(self.h5file, TRIGGER_TABLE)
+
+        self.instrument_table = None
+        if self.load_instrument:
+            table = self.subarray.to_table()
+            optics = self.subarray.to_table(kind="optics")
+            optics["optics_index"] = np.arange(len(optics))
+            optics.remove_columns(["name", "description", "type"])
+            table = join(
+                table,
+                optics,
+                keys="optics_index",
+                # conflicts for TAB_VER, TAB_TYPE, not needed here, ignore
+                metadata_conflicts="silent",
+            )
+
+            table.remove_columns(["optics_index", "camera_index"])
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _get_reconstructors(self):
+
+        if GEOMETRY_GROUP in self.h5file:
+            g = self.h5file.root[GEOMETRY_GROUP]
+        else:
+            raise ValueError(f"No DL2 geometry data in h5file: {self.h5file}")
+        
+        reconstructors = g._v_children.keys()
+        
+        return reconstructors
+
+    def _read_telescope_table(self, group, tel_id):
+
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+    
+    def _read_telescope_DL2_geometry(self, group, reconstructor):
+        
+        key = f"{group}/{reconstructor}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=None)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, labels=None, sort_by=["obs_id", "event_id"]):
+        """Read data from file.
+        
+        Parameters
+        ----------
+        labels: list
+            Either tel_ids or tel_types depending on dataset split.
```

Comment:
Why use a double meaning here? For the telescope events we have `by_telescope_type` and `by_telescope_id`.

Also, why is this needed for *subarray* events?

Reply:
because this function does all, that's why it was called more generically `read_data`
you can provide either tel_ids or tel_types regardless of the input file dataset splitting mode

you do not need to use any of the specific functions by*** (which in fact we could make private)

Reply:
That's not good. We should have a clear api without these double meanings.

Reply:
I should be able to split it easily in 2, but I don't see the problem honestly 

you write in the docstring parameters "Either list of tel_ids or list of tel_types" and similar for the returns...

Reply:
So if this is basically just `read_telescope_events_by_id` if labels is a list of ids and `read_telescope_events_by_type` if `labels` is a list of types, why is this function so complicated???

Reply:
I wrote a proper docstring, it should be fool-proof now

Reply:
> So if this is basically just `read_telescope_events_by_id` if labels is a list of ids and `read_telescope_events_by_type` if `labels` is a list of types

No, because it also accounts for the case in which I read from a file (regardless of how it has been split at writing time) and I provide a list of tel_ids that span different tel_types and I still want them to result in tel-type based tables

Reply:
reset to old commit - resolving